src/lib/ares_process.c

```diff
@@ -817,6 +848,76 @@ static struct server_state *ares__random_server(ares_channel_t *channel)
   return NULL;
 }
 
+/* Pick a priority server from the list.
+ *
+ * We default to using the first server in the sorted list of servers. That is
+ * the server with the lowest number of consecutive failures, and then the
+ * highest priority server (by idx) if there is a draw.
+ *
+ * However, this by itself has two issues:
+ * (1) If the highest priority server temporarily goes down and hits
+ *     some failures, then that server will never be retried until all other
+ *     servers hit the same number of failures. This may prevent the
+ *     highest priority server from being retried for a long time.
+ * (2) If all servers are unhealthy, then we will continue to select the
+ *     server with the lowest number of consecutive failures even though it may
+ *     be more broken than other servers later in the list.
+ *
+ * For issue (1), if the first server in the sorted list is not the highest
+ * priority server, then with some probability we select a higher priority
+ * server at random instead.
+ *
+ * For issue (2), if all servers in the list are unhealthy (and not in the
+ * process of recovering), then we fall back to selecting a random server.
+ */
+static struct server_state *ares__priority_server(ares_channel_t *channel)
+{
+  struct server_state *server;
+  unsigned char        c;
+  size_t               idx;
+  ares__slist_node_t  *node;
+  struct server_state *node_val;
+
+  /* Select the best choice of server, based on consecutive failures and
+   * server idx.
+   */
+  server = ares__slist_first_val(channel->servers);
+
+  /* Defensive code against no servers being available on the channel. */
+  if (server == NULL) {
+    return NULL;
+  }
+
+  if (server->consec_failures >= channel->server_failure_threshold) {
+    /* All servers are unhealthy so fall back to a random server. */
+    return ares__random_server(channel);
+  }
+
+  if (server->idx > 0) {
+    /* The best server is not the highest priority server. To avoid never
+     * retrying higher priority servers, with some probability we select a
+     * higher priority server at random instead.
+     */
+    ares__rand_bytes(channel->rand_state, &c, 1);
+    if ((size_t)c % channel->priority_server_chance == 0) {
```

Comment:
if `channel->priority_server_chance` is never initialized, this is a divide by zero.

Reply:
Makes sense, I will add a check that `channel->priority_server_chance > 0` (otherwise we can just go for the first server in the list).